File: storage/db-core/src/main/java/com/coffee/gu/BaseCustomIdEntity.java
package com.coffee.gu;

import jakarta.persistence.*;
import org.springframework.data.domain.Persistable;

@MappedSuperclass
public abstract class BaseCustomIdEntity<T> extends BaseTimeEntity implements Persistable<T> {

    @Id
    protected T id;
    @Transient
    private boolean isNewEntity = true;

    protected BaseCustomIdEntity() {}

    protected BaseCustomIdEntity(T id, boolean isNewEntity) {
        this.id = id;
        this.isNewEntity = isNewEntity;
    }

    @Override
    public T getId() {
        return this.id;
    }

    @Override
    public boolean isNew() {
        return this.isNewEntity;
    }

    @PostLoad
    @PostPersist
    protected void markNotNew() {
        this.isNewEntity = false;
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/BaseCustomIdEntity.kt
package com.coffee.gu

import jakarta.persistence.MappedSuperclass
import jakarta.persistence.PostLoad
import jakarta.persistence.PostPersist
import jakarta.persistence.Transient
import org.springframework.data.domain.Persistable

@MappedSuperclass
abstract class BaseCustomIdEntity<T : Any>(
    @Transient
    var isNewEntity: Boolean = true,
) : BaseEntity(), Persistable<T> {

    abstract override fun getId(): T

    override fun isNew(): Boolean = this.isNewEntity

    @PostLoad
    @PostPersist
    protected fun markNotNew() {
        this.isNewEntity = false
    }
}
